Add CSRF token prevention.

cmeiklejohn opened the pull request:
Add CSRF token prevention to Riak Control.
- Break out resources which perform potentially destructive actions.
- Switch these from GET requests to POST requests.
- No longer serve index.html off of the file system, serve from erlydtl template so we can insert a CSRF token into the header as a meta tag. (Potentially problematic because users might have bookmarked /admin/ui/index.html).

The changes to the API are less than awesome, but are good enough to get us to a place where things feel much safter (POST vs non-idempotent GET).

cmeiklejohn committed:
Add CSRF token prevention.

cmeiklejohn committed:
Typo correction.

cmeiklejohn commented:
Ah, cool.  @dizzyd pointed me at rel/reltool.config in basho/riak.  Once I get this working, I'll open a pull request there.

cmeiklejohn committed:
Ensure page is returned as text/html.

cmeiklejohn committed:
Resolve missing semi, ensure URI encoding.

seancribbs commented on src/admin_cluster_down.erl:
```diff
@@ -0,0 +1,73 @@
+    [{admin_routes:cluster_route(["down",node]),?MODULE,[]}].
+
+%% entry-point for the resource from webmachine
+init (Action) ->
```
What is `Action` here? Should be just an empty list, right? Seems unnecessary to pass it along in the resource state.

cmeiklejohn replied:
Corrected in [0ec17aa3624a20eb81df6a5929707d0b17677e2b].

seancribbs commented on src/admin_cluster_join.erl:
```diff
@@ -0,0 +1,85 @@
+
+%% defines the webmachine routes this module handles
+routes () ->
+    [{admin_routes:cluster_route(["join",node]),?MODULE,[]}].
```
Out of curiosity, in this route and the one for the 'down' command, why is the node name in the URL and not part of the POST?

cmeiklejohn replied:
Ah, all of these destructive functions in the API (see cluster join, leave, node down, stop) put the node name in the URI.  I can work on changing that; what are your thoughts on addressing that in a different pull request?

seancribbs replied:
Seems appropriate not to grow the scope of this fix, defer.

seancribbs commented on src/admin_gui.erl:
```diff
@@ -52,20 +49,6 @@ routes () ->
 
-%% redirect to the index page if no file given
-moved_permanently (Req,index) ->
-    {{true,"/admin/ui/index.html"},Req,undefined};
```
Rather than removing it completely, you might want to redirect the other way; that is, if they request `index.html`, point to the template-rendered resource instead.

cmeiklejohn replied:
Corrected in [21e15721b875a0b98570f6e53b20b480b3d992df].

cmeiklejohn committed:
Whitespace correction.

cmeiklejohn committed:
No longer pass empty state around.

cmeiklejohn committed:
Redirect old index to the new template.

cmeiklejohn committed:
Rename method.

cmeiklejohn committed:
Add null origin check.

cmeiklejohn committed:
Validate origin.

cmeiklejohn committed:
Break fallbacks out and use origin checking.

cmeiklejohn committed:
Remove content_provided call and check origin.

cmeiklejohn committed:
Fix bugs with output formatting.

Fix numerous bugs where the input to the result 
function was not uniform and would throw 500 
errors.

cmeiklejohn committed:
Return generic error for now.

This never returned anything other than Internal Server Error, so
I'm just hardcoding the string to return an error until we get code
which can parse the body.

cmeiklejohn commented:
See: basho/riak_ee#73, basho/riak#182.

coderoshi commented on priv/admin/js/cluster.js:
```diff
@@ -139,15 +139,22 @@ $(document).ready(function () {
-            dataType: 'json',
+            type: 'POST',
+            data: "csrf_token=" + encodeURIComponent(csrf_token),
+            dataType: 'application/json',
```
I'm pretty sure it was correct the first time as "dataType: 'json'"

cmeiklejohn replied:
You are absolutely right; nice catch!  Fixed in [08e3f2a7410adce7dff41b54cb7faca019a50fdd].

cmeiklejohn committed:
Supply correct content type.

coderoshi commented on src/riak_control_security.erl:
```diff
@@ -126,3 +129,38 @@ valid_userpass(_User, _Pass, _Auth) ->
+            true;
+        _ ->
+            false
+    end.
```
I'm not much of an erlang programmer, but doesn't _ -> include undefined? Wouldn't this suffice?

CookieToken ->
    true;
_ ->
    false

cmeiklejohn replied:
The problem is when CookieToken and BodyToken are both undefined, the first clause matches.

coderoshi replied:
gotcha - I'm not used to the "capitalization equals variable" thing

seancribbs replied:
@cmeiklejohn Sure, but he's still right. You're not doing anything special with the `undefined` clause. Why not simplify this case clause to:

``` erlang
BodyToken /= undefined andalso BodyToken == CookieToken.
```

cmeiklejohn replied:
Very nice, thanks.  I think this is covered in [21d8b19bf0d5891fa1a5a8b1f1f040f5058bd9a9].

seancribbs commented:
Other than the case-clause wonkiness, I'm +1 to this. Make it happen.

cmeiklejohn committed:
Clean up case statement.
